packages/vjudge/src/providers/yacs.ts

```diff
@@ -0,0 +1,172 @@
+    }
+
+    token: string;
+
+    get loggedIn(): Promise<boolean> {
+        return new Promise((resolve) => {
+            this.get('/user/fetchByToken').query({ token: this.token })
+                .end((err, { status }) => resolve(status === 200));
+        });
+    }
+
+    async ensureLogin() {
+        if (await this.loggedIn) return true;
+        logger.info('retry login');
+        try {
+            // NOTE: you should pass a pre-hashed key!
+            const { body: { token } } = await this.post('/user/login')
+                .send({ username: this.account.handle, password: this.account.password });
+            this.token = token;
+        } catch (e) { }
+        return await this.loggedIn;
+    }
+
+    async getProblem(id: string) {
+        logger.info(id);
+        const { problem } = await this.getData(`/problem/${id.split('P')[1]}`);
+        const tag = [];
+        if (problem.level) tag.push(problem.level);
+        if (problem.contest) tag.push(problem.contest.name);
+        let content = '';
+        if (problem.background.trim()) content += `## 题目背景\n\n${problem.background.trim()}\n\n`;
+        content += `## 题目描述\n\n${problem.description.trim()}\n\n`;
+        content += `## 输入格式\n\n${problem.inputFormat.trim()}\n\n`;
+        content += `## 输出格式\n\n${problem.outputFormat.trim()}\n\n`;
+        content += problem.exampleList.map((sample, index) => {
+            const sampleId = index + 1;
+            let ret = '';
+            ret += `\`\`\`input${sampleId}\n${sample.input.trim()}\n\`\`\`\n\n`;
+            ret += `\`\`\`output${sampleId}\n${sample.output.trim()}\n\`\`\`\n\n`;
+            if (sample.note && sample.note.trim()) ret += `## 样例解释 ${sampleId}\n\n${sample.note.trim()}\n\n`;
+            return ret;
+        }).join('');
+        content += `## 数据范围\n\n${problem.dataRange.trim()}\n\n`;
+        if (problem.source.trim()) content += `## 来源\n\n${problem.source.trim()}\n\n`;
+        return {
+            title: problem.title,
+            data: {
+                'config.yaml': Buffer.from(yaml.dump({
+                    time: `${problem.limitTime}ms`,
+                    memory: `${problem.limitMemory}m`,
+                    type: 'remote_judge',
+                    subType: 'yacs',
+                    target: id,
+                })),
+            },
+            files: {},
+            tag,
+            content,
+        };
+    }
+
+    async listProblem(page: number, resync = false) {
+        if (resync && page > 1) return [];
+        const data = await this.getData(`/problem?pi=${page}`);
+        return data.problemList
+            .filter((problem) => problem.contest.status === '榜单已公布')
+            .map((problem) => `P${problem.id}`);
+    }
+
+    async submitProblem(id: string, lang: string, code: string, info, next, end) {
+        const langs = {
+            'yacs.1': 'C++',
+            'yacs.2': 'Python 3.6',
```

Comment:
using a seperate scope is no longer recommended. please use `validAs` property.

https://github.com/hydro-dev/Hydro/blob/master/packages/vjudge/src/model.ts#L50

```
langs:
  cc:
    compile: g++ ...
    execute: /w/code
    validAs:
      yacs: C++
    ...
```

while the new format provides additional features like pretest.